controllers/etcd_copy_backups_controller.go

```diff
@@ -0,0 +1,409 @@
+				Kind:     "Job",
+				APIGroup: pointer.StringPtr(batchv1.SchemeGroupVersion.String()),
+			}
+			etcdCopyBackupsTask.Status.LastOperation = lastOperation
+		}
+		etcdCopyBackupsTask.Status.ObservedGeneration = &etcdCopyBackupsTask.Generation
+		return nil
+	})
+}
+
+func (r *EtcdCopyBackupsTaskReconciler) getLastOperation(ctx context.Context, etcdCopyBackupsTask *druidv1alpha1.EtcdCopyBackupsTask) (*druidv1alpha1.LastOperation, error) {
+	lastOperation := &druidv1alpha1.LastOperation{}
+	job := &batchv1.Job{}
+
+	if err := r.Get(ctx, kutil.Key(etcdCopyBackupsTask.Namespace, etcdCopyBackupsTask.Name+workerSuffix), job); err != nil {
+		return nil, err
+	}
+
+	if job.Status.Active > 0 {
+		lastOperation.State = druidv1alpha1.LastOperationStateActive
+		lastOperation.Description = fmt.Sprintf("Job %s is still active", job.Name)
+	}
+	if job.Status.Succeeded > 0 {
+		lastOperation.State = druidv1alpha1.LastOperationStateSucceeded
+		lastOperation.Description = fmt.Sprintf("Job %s succeeded", job.Name)
+	}
+	if job.Status.Failed > 0 {
+		lastOperation.State = druidv1alpha1.LastOperationStateFailed
+		lastOperation.Description = fmt.Sprintf("Job %s failed", job.Name)
+	}
+	lastOperation.LastUpdateTime = v1.Now()
+
+	return lastOperation, nil
+}
+
+func (r *EtcdCopyBackupsTaskReconciler) removeOperationAnnotation(ctx context.Context, logger logr.Logger, etcdCopyBackupsTask *druidv1alpha1.EtcdCopyBackupsTask) error {
+	if _, ok := etcdCopyBackupsTask.Annotations[v1beta1constants.GardenerOperation]; ok {
+		objectCopy := etcdCopyBackupsTask.DeepCopy()
+		annotations := objectCopy.GetAnnotations()
+		delete(annotations, v1beta1constants.GardenerOperation)
+		objectCopy.SetAnnotations(annotations)
+		if err := r.Patch(ctx, objectCopy, client.MergeFrom(etcdCopyBackupsTask)); client.IgnoreNotFound(err) != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *EtcdCopyBackupsTaskReconciler) getCopyJob(etcdCopyBackupsTask *druidv1alpha1.EtcdCopyBackupsTask, values map[string]interface{}, logger logr.Logger) (*batchv1.Job, error) {
+	decoded := &batchv1.Job{}
+	copyJobPath := filepath.Join("etcd", "templates", "etcd-copy-backups-job.yaml")
+	chartPath := getChartPath()
+
+	renderedChart, err := r.chartApplier.Render(chartPath, etcdCopyBackupsTask.Name, etcdCopyBackupsTask.Namespace, values)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := decodeObject(renderedChart, copyJobPath, &decoded); err != nil {
+		return nil, err
+	}
+
+	return decoded, nil
+}
+
+func (r *EtcdCopyBackupsTaskReconciler) runCopyJob(ctx context.Context, etcdCopyBackupsTask *druidv1alpha1.EtcdCopyBackupsTask, values map[string]interface{}, logger logr.Logger) (*batchv1.Job, error) {
+	job, err := r.getCopyJob(etcdCopyBackupsTask, values, logger)
+	if err != nil {
+		return nil, err
+	}
+	return job, r.Create(ctx, job)
+}
+
+func (r *EtcdCopyBackupsTaskReconciler) getValuesMap(etcdCopyBackupsTask *druidv1alpha1.EtcdCopyBackupsTask) (map[string]interface{}, error) {
```

Comment:
The only think you need from `EtcdCopyBackupsTaskReconciler` is the `imageVector`. My suggestion: Remove the pointer receiver, pass `imageVector` as a function argument. Then you can easily add unit tests.

Reply:
I tend to disagree with this suggestion. I would rather not make a method that naturally belongs to the type a standalone function just because it's a bit easier to add unit tests. Also, it's a unexported method that from my perspective shouldn't be tested on its own. I see we already have unit tests for `etcd_controller.go`, shouldn't we adopt a similar approach for this controller?

Reply:
Also fine for me 👍